dali/python/nvidia/dali/plugin/pytorch/__init__.py

```diff
@@ -795,3 +853,266 @@ def __next__(self) -> List[Dict[str, torch.Tensor]]:
+        """
+        def __init__(self, dali_proxy, inputs):
+            self.dali_proxy = dali_proxy
+            self.inputs = inputs
+            assert len(self.inputs) == dali_proxy.num_inputs
+
+    def transform(self, *inputs):
+        """
+        The 'transform' function consists of returning a reference to the pipeline run
+        """
+        assert len(inputs) == self.num_inputs, f"Unexpected number of inputs: {len(inputs)}"
+        return DALIProxy.PipelineRunRef(self, inputs)
+
+
+def collate_pipeline_run_ref_fn(pipe_out, *, collate_fn_map=None):
+    """
+    Special collate function that schedules a batch for execution
+    """
+    assert len(pipe_out) > 0
+    first_elem = pipe_out[0]
+    inputs = [[] for idx in range(len(first_elem.inputs))]
+    for elem in pipe_out:
+        assert first_elem.dali_proxy == elem.dali_proxy
+        for idx, input_ref in enumerate(elem.inputs):
+            inputs[idx].append(input_ref)
+    return first_elem.dali_proxy.schedule_batch(inputs)
+
+
+def custom_collate(batch):
+    """
+    Subscribe a special collate function for PipelineRunRef, that handles the scheduling of the iteration
+    on the fly
+    """
+    collate_fn_map = data._utils.collate.default_collate_fn_map
+    collate_fn_map.update({DALIProxy.PipelineRunRef: collate_pipeline_run_ref_fn})
+    return collate(batch, collate_fn_map=collate_fn_map)
+
+
+def flatten_tuple(nested_tuple):
+    """
+    Flattens a nested tuple
+    """
+    flat_list = []
+
+    def _flatten(t):
+        for item in t:
+            if isinstance(item, tuple):
+                _flatten(item)
+            else:
+                flat_list.append(item)
+
+    _flatten(nested_tuple)
+    return tuple(flat_list)
+
+
+class DALIMultiProcessingDataLoaderIter(_MultiProcessingDataLoaderIter):
+    """
+    Data loader iterator used by the DALI proxy data loader
+    """
+    def __init__(self, loader):
+        super().__init__(loader)
+        self.loader = loader
+
+    def _next_data(self):
+        data = super()._next_data()
+        if not hasattr(data, "__iter__"):
+            print(
+                "Warning: Non iterable returned from dataloader. Please "
+                " review the code, since it usually indicates a bug in a pipeline."
+            )
+            data = [data]
+        for data_idx, data_elem in enumerate(data):
+            # If loader returns a dictionary the iterator iterates over its keys.
+            # We need to access a value. Probably need to address more casess.
+            if isinstance(data, dict):
+                if isinstance(data[data_elem], DALIProxy.PipelineOutputRef):
+                    data[data_elem] = self.loader.get_outputs(data[data_elem].info)
+            if isinstance(data_elem, DALIProxy.PipelineOutputRef):
+                data[data_idx] = self.loader.get_outputs(data_elem.info)
+        return data
+
+
+class DALIDataLoader(DataLoader):
```

Comment:
This name was good in the prototyping stage, but I believe we should now change it to indicate what it actually does. Or, if we decide to extract Proxy to a submodule, it can be just `dali.plugin.torch.proxy.DataLoader`

Reply:
Done